hey, i pulled your branch and ran the build and i got some errors:

```
drivers/i2c/busses/i2c-bl808.c: In function 'bl808_i2c_disable':
drivers/i2c/busses/i2c-bl808.c:428:9: error: implicit declaration of function 'bl808_i2c_clear_fifo_err' [-Werror=implicit-function-declaration]
  428 |         bl808_i2c_clear_fifo_err(i2c_dev);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
drivers/i2c/busses/i2c-bl808.c:429:9: error: implicit declaration of function 'bl808_i2c_clear_interrupts' [-Werror=implicit-function-declaration]
  429 |         bl808_i2c_clear_interrupts(i2c_dev);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
drivers/i2c/busses/i2c-bl808.c: At top level:
drivers/i2c/busses/i2c-bl808.c:492:13: warning: conflicting types for 'bl808_i2c_clear_interrupts'
  492 | static void bl808_i2c_clear_interrupts(struct bl808_i2c_dev *i2c_dev) {
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~
drivers/i2c/busses/i2c-bl808.c:492:13: error: static declaration of 'bl808_i2c_clear_interrupts' follows non-static declaration
drivers/i2c/busses/i2c-bl808.c:429:9: note: previous implicit declaration of 'bl808_i2c_clear_interrupts' was here
  429 |         bl808_i2c_clear_interrupts(i2c_dev);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
drivers/i2c/busses/i2c-bl808.c:500:13: warning: conflicting types for 'bl808_i2c_clear_fifo_err'
  500 | static void bl808_i2c_clear_fifo_err(struct bl808_i2c_dev *i2c_dev) {
      |             ^~~~~~~~~~~~~~~~~~~~~~~~
drivers/i2c/busses/i2c-bl808.c:500:13: error: static declaration of 'bl808_i2c_clear_fifo_err' follows non-static declaration
drivers/i2c/busses/i2c-bl808.c:428:9: note: previous implicit declaration of 'bl808_i2c_clear_fifo_err' was here
  428 |         bl808_i2c_clear_fifo_err(i2c_dev);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/drivers/i2c/busses/i2c-bl808.c
+++ b/drivers/i2c/busses/i2c-bl808.c
@@ -417,23 +417,30 @@ static void bl808_i2c_enable(struct bl808_i2c_dev *i2c_dev){
         bl808_i2c_writel(i2c_dev, BL808_I2C_CONFIG, val);
 }
 
+static void bl808_i2c_clear_interrupts(struct bl808_i2c_dev *i2c_dev) {
+        u32 val = BL808_I2C_STS_END_CLR |
+                  BL808_I2C_STS_NAK_CLR |
+                  BL808_I2C_STS_ARB_CLR;
+
+        bl808_i2c_writel(i2c_dev, BL808_I2C_STS, val);
+}
+
+static void bl808_i2c_clear_fifo_err(struct bl808_i2c_dev *i2c_dev) {
+        u32 val = (BL808_I2C_FIFO_CONFIG_0_RX_FIFO_CLR |
+                   BL808_I2C_FIFO_CONFIG_0_TX_FIFO_CLR);
+
+        bl808_i2c_writel(i2c_dev,  BL808_I2C_FIFO_CONFIG_0, val);
+}
+
 static void bl808_i2c_disable(struct bl808_i2c_dev *i2c_dev){
         u32 val;
         /* disable i2c */
         val = bl808_i2c_readl(i2c_dev, BL808_I2C_CONFIG);
         val &= ~BL808_I2C_CONFIG_M_EN;
         bl808_i2c_writel(i2c_dev, BL808_I2C_CONFIG, val);
-        /* Clear I2C fifo */
-        val = bl808_i2c_readl(i2c_dev, BL808_I2C_FIFO_CONFIG_0);
-        val |= BL808_I2C_FIFO_CONFIG_0_TX_FIFO_CLR;
-        val |= BL808_I2C_FIFO_CONFIG_0_RX_FIFO_CLR;
-        bl808_i2c_writel(i2c_dev, BL808_I2C_FIFO_CONFIG_0, val);
-        /* Clear I2C interrupt status */
-        val = bl808_i2c_readl(i2c_dev, BL808_I2C_STS);
-        val |= BL808_I2C_STS_END_CLR;
-        val |= BL808_I2C_STS_NAK_CLR;
-        val |= BL808_I2C_STS_ARB_CLR;
-        bl808_i2c_writel(i2c_dev, BL808_I2C_STS, val);
+
+        bl808_i2c_clear_fifo_err(i2c_dev);
+        bl808_i2c_clear_interrupts(i2c_dev);
 }
 
 static void bl808_i2c_enable_interrupts(struct bl808_i2c_dev *i2c_dev, u32 interrupts) {
@@ -494,29 +501,6 @@ static void bl808_i2c_mask_interrupts(struct bl808_i2c_dev *i2c_dev, u32 interru
                 BL808_I2C_STS_FER_MASK));
 
         bl808_i2c_writel(i2c_dev, BL808_I2C_STS, val);
-}
-
-static void bl808_i2c_clear_interrupts(struct bl808_i2c_dev *i2c_dev) {
-        u32 val;
-
-        val = bl808_i2c_readl(i2c_dev, BL808_I2C_STS);
-
-        val |= (BL808_I2C_STS_END_CLR |
-                BL808_I2C_STS_NAK_CLR |
-                BL808_I2C_STS_ARB_CLR);
-
-        bl808_i2c_writel(i2c_dev, BL808_I2C_STS, val);
-}
-
-static void bl808_i2c_clear_fifo_err(struct bl808_i2c_dev *i2c_dev) {
-        u32 val;
-
-        val = bl808_i2c_readl(i2c_dev, BL808_I2C_FIFO_CONFIG_0);
-
-        val |= (BL808_I2C_FIFO_CONFIG_0_RX_FIFO_CLR |
-                BL808_I2C_FIFO_CONFIG_0_TX_FIFO_CLR);
-
-        bl808_i2c_writel(i2c_dev,  BL808_I2C_FIFO_CONFIG_0, val);
 }
 
 static void bl808_i2c_init(struct bl808_i2c_dev *i2c_dev) {

--- a/drivers/i2c/busses/i2c-bl808.c
+++ b/drivers/i2c/busses/i2c-bl808.c
@@ -314,7 +314,7 @@ static void bl808_fill_tx_fifo(struct bl808_i2c_dev *i2c_dev) {
 
         tx_fifo_free = (val & BL808_I2C_FIFO_CONFIG_1_TX_FIFO_CNT_MASK) >> BL808_I2C_FIFO_CONFIG_1_TX_FIFO_CNT_SHIFT;
 
-        while (tx_fifo_free > 0) {
+        while (tx_fifo_free > 0 && i2c_dev->msg_buf_remaining > 0) {
                 u16 bytes_to_fill = min_t(u16, i2c_dev->msg_buf_remaining, 4);
                 u32 temp = 0;
 

--- a/drivers/i2c/busses/i2c-bl808.c
+++ b/drivers/i2c/busses/i2c-bl808.c
@@ -308,7 +308,6 @@ static struct clk *bl808_i2c_register_div(struct device *dev, struct clk *mclk, 
 
 static void bl808_fill_tx_fifo(struct bl808_i2c_dev *i2c_dev) {
         u32 val;
-        u32 temp = 0;
         u32 tx_fifo_free;
 
         val = bl808_i2c_readl(i2c_dev, BL808_I2C_FIFO_CONFIG_1);
@@ -316,23 +315,14 @@ static void bl808_fill_tx_fifo(struct bl808_i2c_dev *i2c_dev) {
         tx_fifo_free = (val & BL808_I2C_FIFO_CONFIG_1_TX_FIFO_CNT_MASK) >> BL808_I2C_FIFO_CONFIG_1_TX_FIFO_CNT_SHIFT;
 
         while (tx_fifo_free > 0) {
+                u16 bytes_to_fill = min_t(u16, i2c_dev->msg_buf_remaining, 4);
+                u32 temp = 0;
 
-                temp = 0;
+                for (u8 i = 0; i < bytes_to_fill; i++)
+                        temp += i2c_dev->msg_buf[i] << ((i % 4) * 8);
+                i2c_dev->msg_buf += bytes_to_fill;
+                i2c_dev->msg_buf_remaining -= bytes_to_fill;
 
-                if (i2c_dev->msg_buf_remaining >= 4) {
-                        for(u8 i = 0; i < 4; i++) {
-                                temp += i2c_dev->msg_buf[i] << ((i % 4) * 8);
-                        }
-                        i2c_dev->msg_buf += 4;
-                        i2c_dev->msg_buf_remaining -= 4;
-                } else if (i2c_dev->msg_buf_remaining > 0) {
-                        for(u8 i = 0; i < i2c_dev->msg_buf_remaining; i++) {
-                                temp += i2c_dev->msg_buf[i] << ((i % 4) * 8);
-                        }
-                        i2c_dev->msg_buf += i2c_dev->msg_buf_remaining;
-                        i2c_dev->msg_buf_remaining -= i2c_dev->msg_buf_remaining;
-                }
-                
                 bl808_i2c_writel(i2c_dev, BL808_I2C_FIFO_WDATA, temp);
 
                 val = bl808_i2c_readl(i2c_dev, BL808_I2C_FIFO_CONFIG_1);

